Make pyodbc.Row and databricks.Row JSON-serializable via new `make_serializable` method

Joffreybvn opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
The ODBCHook returns pyodbc.Row objects when used with SQLExecuteQueryOperator to do SELECT queries, which cause serialization errors in the XCom backend. This PR follows the [discussion started here](https://github.com/apache/airflow/discussions/32058#discussioncomment-6263179).


A good place to implement the transformation of Row into tuples is after the execution of the handler, in the [`run()`](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/hooks/sql.py#L355) method. There, the raw data structure is available. Doing it later would imply to deal with potential nested structure (ex: list of results).

Thus, I propose to add method in the DBApiHook to make the result of a query serializable. So that subclasses can override it to implement their own logic.

I see that many Hooks based on `DbApiHook` have a custom run method. But considering @potiuk comment about "[the most standard the better](https://github.com/apache/airflow/discussions/32058#discussioncomment-6263179)", I propose to go for an extra internal method that Hooks can override, rather than copying the full run() method just to implement a small change. Considering also that the pyodbc.Row may not be the only case where a custom object causes issues ([Databricks could be rewrote to handle that issue](https://github.com/apache/airflow/pull/31780/files) at Hook level). And considering that this pattern already exists with the [`serialize_cell`](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/hooks/sql.py#L478) method which get overidden by child hooks.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

joffreybienvenu-infrabel committed:
feat: Transform pyodbc.Row objects into tuples in Hook.run() method

joffreybienvenu-infrabel committed:
feat: Transform pyodbc.Row objects into tuples in Hook.run() method

joffreybienvenu-infrabel committed:
feat: Transform pyodbc.Row objects into tuples in Hook.run() method

joffreybienvenu-infrabel committed:
feat: Add serialization test to ODBCHook

joffreybienvenu-infrabel committed:
feat: Add serialization test to ODBCHook

joffreybienvenu-infrabel committed:
feat: Add serialization test to ODBCHook

utkarsharma2 commented on airflow/providers/odbc/hooks/odbc.py:
```diff
@@ -211,3 +211,10 @@ def get_sqlalchemy_connection(
+        """Transform the pyodbc.Row objects returned from a SQL command into
+        JSON-serializable objects.
+        """
+        return [tuple(row) for row in result]
```
I think this would be a breaking change for existing users, they would expect the return value to be `list[pyodbc.Row]` and not `list[tuple]`. Maybe we can change the type based on a bool flag?

Joffreybvn replied:
Indeed, I added a flag which disable the serializable data structure by default.

joffreybienvenu-infrabel committed:
feat: Add `return_serializable` flag with default to False

joffreybienvenu-infrabel committed:
feat: Add `return_serializable` flag with default to False

joffreybienvenu-infrabel committed:
fix: Update test with `return_serializable` parameter

joffreybienvenu-infrabel committed:
fix: Update test with `return_serializable` parameter

joffreybienvenu-infrabel committed:
fix: Make superclass `_make_serializable` compatible with subclass

utkarsharma2 commented on airflow/providers/odbc/hooks/odbc.py:
```diff
@@ -65,6 +67,7 @@ def __init__(
         dsn: str | None = None,
         connect_kwargs: dict | None = None,
         sqlalchemy_scheme: str | None = None,
+        return_serializable: bool | None = False,
```
```suggestion
        return_serializable: bool = False,
```

utkarsharma2 commented on tests/providers/odbc/hooks/test_odbc.py:
```diff
@@ -19,31 +19,55 @@
     def get_hook(self=None, hook_params=None, conn_params=None):
         hook_params = hook_params or {}
         conn_params = conn_params or {}
+
```
```suggestion
```

utkarsharma2 commented on tests/providers/odbc/hooks/test_odbc.py:
```diff
@@ -235,3 +259,15 @@ def test_sqlalchemy_scheme_extra(self):
         uri = hook.get_uri()
         assert urlsplit(uri).scheme == "my-scheme"
+
+    def test_query_return_value_serialization(self, Row):
```
Maybe we can also add a testcase for `return_serializable=False`?

Joffreybvn replied:
Added it

Lee-W commented on airflow/providers/odbc/hooks/odbc.py:
```diff
@@ -211,3 +215,9 @@ def get_sqlalchemy_connection(
         cnx = engine.connect(**(connect_kwargs or {}))
         return cnx
+
+    def _make_serializable(self, result: list[pyodbc.Row]) -> list[tuple] | list[pyodbc.Row]:
```
nitpick: Should the return type be annotated this one?

```suggestion
    def _make_serializable(self, result: list[pyodbc.Row]) -> list[tuple[pyodbc.Row]] | list[pyodbc.Row]:
```

Joffreybvn replied:
No, when the result is made serializable, Row objects are transformed into pure Python data structure (tuples). Thus, in this case, it should either return a list of tuple, or return directly the original data.

Lee-W commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -372,6 +372,15 @@ def run(
+
+        This method is intended to be overridden by subclasses of the
+        `DbApiHook`. Its purpose is to transform the result of a SQL command
+        into a JSON-serializable format.
```
What should happen if this is not overridden? If we don't want the user do not override it, should we add an exception here? If it's ok that we return the original input as output, could we add this behavior to the docstring?

Joffreybvn replied:
In most cases, this won't be overidden, and indeed the original input is returned as output. For what I know, the only cases where they are custom Row objects are in Databricks and PyODBC. I'll edit the docstring.

Joffreybvn committed:
fix: Remove optional None type for return_serilizable

Co-authored-by: Utkarsh Sharma <utkarsharma2@gmail.com>

Joffreybvn committed:
fix: Add space to test case

Co-authored-by: Utkarsh Sharma <utkarsharma2@gmail.com>

joffreybienvenu-infrabel committed:
feat: Add test for return_serializable to False

joffreybienvenu-infrabel committed:
feat: Document default behavior of `_make_serializable` method

joffreybienvenu-infrabel committed:
feat: Make pyodbc.Row serializable on ODBCHook via handler injection

joffreybienvenu-infrabel committed:
feat: Make pyodbc.Row serializable on ODBCHook via handler injection

joffreybienvenu-infrabel committed:
feat: Implement pyodbc.Row to typing.Namedtuple conversion for ODBCHook

joffreybienvenu-infrabel committed:
feat: Move `_make_serializable` method into common.sql

joffreybienvenu-infrabel committed:
feat: Re-implement Databricks serialization at hook level

joffreybienvenu-infrabel committed:
feat: Add static check for SQL providers which depends on make_serializable

joffreybienvenu-infrabel committed:
feat: Add note for maintainers

joffreybienvenu-infrabel committed:
fix: Bump common-sql dependency in odbc and databricks providers

joffreybienvenu-infrabel committed:
fix: Re-generate static-checks commands

joffreybienvenu-infrabel committed:
fix: bump common.sql to 1.8.1

joffreybienvenu-infrabel committed:
fix: re-generate breeze static-check screenshot

joffreybienvenu-infrabel committed:
fix: ignore mypy static-check error

Instantiating NamedTuple dynamically is not supported by mypy. https://github.com/python/mypy/issues/848

joffreybienvenu-infrabel committed:
fix: re-generate breeze static-check screenshot

potiuk reviewed (approved):
This is really cool. It also seems to be fully backwards-compatible . @utkarsharma2 and @Lee-W - WDYT?

pankajastro commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -240,6 +240,13 @@ def run(
+    def _make_serializable(result):
+        """Transform the databricks Row objects into a JSON-serializable list of rows."""
+        if result is not None:
+            return [list(row) for row in result]
```
Hey @Joffreybvn  I feel If the handler callable returns non-iterable then this would fail. 

I call the run method like below and it's breaking my DAG
```
run( f"SELECT COUNT(*) FROM {table}", handler=lambda x: x.fetchone())
```

**Error**
```
packages/airflow/providers/databricks/hooks/databricks_sql.py", line 247, in <listcomp>
    return [list(row) for row in result]
TypeError: 'int' object is not iterable
```

potiuk replied:
Correct. Unfortunately - the DBAPIHook for it's backwards-compatibility reasons is pretty difficult to handle all cases - including the case where we either return  list of results or one result only.

I tried to explain it in https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md#decision (including all the possible variants of returned values) and I think our `handler` definition is not precise enough to explain that this can happen (we have not been as much into typing as we were back then).

@Joffreybvn -> I think this means we have to hold -on with databricks/pyodbc implementation to add support for those cases. Will you have time to fix it quickly ?
